### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by VaheStepanyan100**
- Ensured that the N+1 problem is solved when fetching all posts and their comments for a user.
 Before:

Screenshot 2023-09-03 at 17 46 29

- Integration specs

- Use Capybara to write integration tests for each view in your project.

- Think on the flow of things and group the tests of page in a way that makes sense. For example, one test would check for the things in the page and another test for clicking links on it.

- User index page:

  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.

After:
 
Screenshot 2023-09-04 at 19 22 34

- user show page:

  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.
- User post index page:

  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.
  
- Post show page:

  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commentor.
  - I can see the comment each commentor left.
  - Be sure to include integration specs for any other features you may have implemented!

### CODE REVIEW

File: spec/features/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    expect(page).to have_content('He is a good programmar')
+    expect(page).not_to have_content('fist text')
+  end
+
+  scenario 'has a link to the user index page' do
+    visit user_path(user)
+    expect(page).to have_button('See all posts')
+    click_link 'See all posts'
+    expect(current_path).to eq(user_posts_path(user))
+  end
```
Comment (brainconnect93) [inline spec/features/user_show_spec.rb]: - [ ] Great effort working on the integration spec of the project milestone, but currently some of the test cases are not passing and require your attention. Checking through the test cases example shown by the `rspec spec/features` command, it shows that you have 12 examples of test cases as against the project requirements. the number of test cases required of you to carry out is around 28-29 test cases. I would kindly suggest you re-visit the project requirement and be sure to have written a test for all the cases and they are all passing. I would advise you double check the test cases you have written and have all the test cases written completely for all the files in your `spec/features`
(i.e `user_show_spec.rb`, `user_index_spec.rb`, `post_show_spec.rb`, & `post_index_spec.rb`).

Images of the test cases requested to test 👇 




You can use the link below as a guideline.
> Note: Do not copy and paste the work in order not to be flagged for plagiarism only use it as a guideline.
https://github.com/DilsherB/ror-blog-app/pull/10/files

**Review by brainconnect93 — CHANGES_REQUESTED** [review]
### Hi @VaheStepanyan100, @Reggeditt, & @DebyGrey, 👋 

# STATUS: Changes Required :recycle:

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Good Points 👍

- Descriptive Pull Request title & summary. ✔️
- Integration specs for views and fix n+1 problems ✔️
- Good use of Git flow ✔️

### Changes Required

- [x]  _Please check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me on( @brainconnect93 ) in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    users
+    visit users_path
+  end
+
+  scenario 'I can see the username, profile picture, and number of posts for each user' do
+    users.each do |user|
+      expect(page).to have_content(user.name)
+      expect(page).to have_content(user.posts_counter)
+    end
+  end
```
Comment (anagudelogu) [inline spec/features/user_index_spec.rb]: - [ ] There are two missing cases here:

image

1. You don't have any expectations for the `user`'s profile picture. 🚧 
2. There's no navigation test. 🚧 

Please consider applying them! 🤓 🔧

File: spec/features/user_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            author: @user1, title: 'Third Post', text: 'My Third post'
+          }
+        ]
+      )
+      page.has_content?(@user1.posts)
+    end
+
+    it 'should have a link to see all posts' do
+      expect(page.html).to include('See all posts')
+    end
```
Comment (anagudelogu) [inline spec/features/user_show_spec.rb]: - [ ] Please make sure to include tests for user interaction and navigation. According to the requirements, these two tests are missing: 🤓 🔧 

image

File: spec/features/post_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      expect(page).to have_content(@comment1.text)
+    end
+
+    it 'should show comments counter' do
+      expect(page).to have_content(@first_post.comments_counter)
+    end
+
+    it 'should show likes counter' do
+      expect(page).to have_content(@first_post.likes_counter)
+    end
```
Comment (anagudelogu) [inline spec/features/post_index_spec.rb]: - [ ] According to the requirements, these tests are missing:

image

Please consider adding them! 🤓 🔧

**Review by anagudelogu — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 🚧 

Hi @VaheStepanyan100 👋🏼 
I'm Andres.

You've done a great job implementing some of the requested changes for this milestone 🚀, but there are still some which haven't been addressed yet. You're getting closer, keep up the good work! 🤖 🎆 



 ## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Descriptive PR title, summary and commit messages. 💬 


 ## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/post_index_spec.rb [hunk 5]
```diff
@@ -1,45 +1,32 @@
...
+  let!(:like1) { Like.create(author: user, post:) }
+
+  scenario "see user's profile picture, username, number of posts and interactions" do
+    visit user_posts_path(user)
+    expect(page).to have_content('Tom')
+    expect(page).to have_css("img[alt='Tom']", count: 1)
+    expect(page).to have_content('1 post')
+    expect(page).to have_content('Comments: 3')
+    expect(page).to have_content('Likes: 1')
+  end
```
Comment (thecodechaser) [inline spec/features/post_index_spec.rb]: - [ ] As requested by the previous reviewer add the below test for the `posts#index` page, You still missed adding that. Please make sure to add that.
image

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @VaheStepanyan100, and @Reggeditt👋, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## To Highlight 🎉
- Good job with integration specs. ✔️
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by julie-ify — APPROVED** [review]
Hi @VaheStepanyan100 and @Reggeditt,

## Highlights!
- Good code organization and working tests ✔️
- Descriptive README ✔️ 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Status: Approved ✔️ 

Cheers and Happy coding!👏👏👏

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**